cdap-app-fabric/src/main/java/co/cask/cdap/route/store/ZKRouteStore.java

```diff
@@ -0,0 +1,214 @@
+    try {
+      return future.get(ZK_TIMEOUT_SECS, TimeUnit.SECONDS);
+    } catch (InterruptedException | ExecutionException | TimeoutException e) {
+      if (e.getCause() instanceof KeeperException.NoNodeException) {
+        throw new NotFoundException(String.format("Route Config for Service %s was not found.", serviceId));
+      }
+      throw Throwables.propagate(e);
+    }
+  }
+
+  private static String getZKPath(ProgramId serviceId) {
+    return String.format("/routestore/%s", ServiceDiscoverable.getName(serviceId));
+  }
+
+  static final Codec<RouteConfig> ROUTE_CONFIG_CODEC = new Codec<RouteConfig>() {
+
+    @Override
+    public byte[] encode(RouteConfig object) throws IOException {
+      return Bytes.toBytes(GSON.toJson(object.getRoutes()));
+    }
+
+    @Override
+    public RouteConfig decode(byte[] data) throws IOException {
+      Map<String, Integer> routes = GSON.fromJson(Bytes.toString(data), MAP_STRING_INTEGER_TYPE);
+      return new RouteConfig(routes);
+    }
+  };
+
+  private Future<RouteConfig> getAndWatchData(final ProgramId serviceId, Watcher watcher) {
+    OperationFuture<NodeData> future = zkClient.getData(getZKPath(serviceId), watcher);
+    final SettableFuture<RouteConfig> settableFuture = routeConfigMap.get(serviceId);
+    Futures.addCallback(future, new FutureCallback<NodeData>() {
+      @Override
+      public void onSuccess(NodeData result) {
+        try {
+          RouteConfig route = ROUTE_CONFIG_CODEC.decode(result.getData());
+          settableFuture.set(route);
+        } catch (IOException ex) {
+          LOG.debug("Unable to deserialize the config for service {}. Got data {}", serviceId, result.getData());
+          // Need to remove the future from the map since later calls will continue to use this future and will think
+          // that there is an exception
+          routeConfigMap.remove(serviceId);
+          settableFuture.setException(ex);
+        }
+      }
+
+      @Override
+      public void onFailure(Throwable t) {
+        routeConfigMap.remove(serviceId);
+        settableFuture.setException(t);
+      }
+    });
+    return settableFuture;
+  }
+
+  @Override
+  public void close() throws Exception {
+    // Clear the map, so that any active watches will expire.
+    routeConfigMap.clear();
+  }
+
+  private class ZKRouteWatcher implements Watcher {
+    private final ProgramId serviceId;
+
+    ZKRouteWatcher(ProgramId serviceId) {
+      this.serviceId = serviceId;
+    }
+
+    @Override
+    public void process(WatchedEvent event) {
+      // If service name doesn't exist in the map, then don't re-watch it.
+      if (!routeConfigMap.containsKey(serviceId)) {
+        return;
+      }
+
+      if (event.getType() == Event.EventType.NodeDeleted) {
+        // Remove the mapping from cache
+        routeConfigMap.remove(serviceId);
+        return;
+      }
+      getAndWatchData(serviceId, this);
```

Comment:
`getAndWatchData()` (called from this watch) sets the settable future in `routeConfigMap` based on the `getData` operation result. But the settable future will already be set by the time this watch is fired. I don't think it is a good idea to set a future multiple times. Instead we could pass a new settable future to `getAndWatchData()` every time it is called. Once the `getData` operation is complete/failed, `getAndWatchData()` will set the future accordingly, and put the future on the `routeConfigMap`